src/game_stats.c

```diff
@@ -0,0 +1,387 @@
+    u32 limit = GameStats_GetMaxValue(statIdx);
+    u32 ret;
+    GameStats_Acquire(gameStats, statIdx);
+    if (value < limit) {
+        ret = GameStats_SetValue(gameStats, statIdx, value);
+    } else {
+        ret = GameStats_SetValue(gameStats, statIdx, limit);
+    }
+    GameStats_Release(gameStats, statIdx);
+    return ret;
+}
+
+u32 GameStats_UpdateBounded(GAME_STATS *gameStats, int statIdx, u32 value) {
+    u32 limit = GameStats_GetMaxValue(statIdx);
+    u32 ret;
+    GameStats_Acquire(gameStats, statIdx);
+    ret = GameStats_GetValue(gameStats, statIdx);
+    if (value > limit) {
+        value = limit;
+    }
+    if (ret < value) {
+        ret = GameStats_SetValue(gameStats, statIdx, value);
+    } else if (ret > limit) {
+        ret = GameStats_SetValue(gameStats, statIdx, limit);
+    }
+    GameStats_Release(gameStats, statIdx);
+    return ret;
+}
+
+u32 GameStats_Inc(GAME_STATS *gameStats, int statIdx) {
+    u32 limit = GameStats_GetMaxValue(statIdx);
+    u32 ret;
+    u32 value;
+    GameStats_Acquire(gameStats, statIdx);
+    value = GameStats_GetValue(gameStats, statIdx) + 1;
+    if (value < limit) {
+        ret = GameStats_SetValue(gameStats, statIdx, value);
+    } else {
+        ret = GameStats_SetValue(gameStats, statIdx, limit);
+    }
+    GameStats_Release(gameStats, statIdx);
+    return ret;
+}
+
+u32 GameStats_Add(GAME_STATS *gameStats, int statIdx, u32 addend) {
+    u32 limit = GameStats_GetMaxValue(statIdx);
+    u32 ret;
+    u32 value;
+    GameStats_Acquire(gameStats, statIdx);
+    value = GameStats_GetValue(gameStats, statIdx);
+    if (value + addend < limit) {
+        ret = GameStats_SetValue(gameStats, statIdx, value + addend);
+    } else {
+        ret = GameStats_SetValue(gameStats, statIdx, limit);
+    }
+    GameStats_Release(gameStats, statIdx);
+    return ret;
+}
+
+u32 GameStats_GetCapped(GAME_STATS *gameStats, int statIdx) {
+    u32 limit = GameStats_GetMaxValue(statIdx);
+    u32 ret;
+    GameStats_Acquire(gameStats, statIdx);
+    ret = GameStats_GetValue(gameStats, statIdx);
+    GameStats_Release(gameStats, statIdx);
+    if (ret > limit) {
+        return limit;
+    } else {
+        return ret;
+    }
+}
+
+u32 GameStats_AddSpecial(GAME_STATS *gameStats, int statIdx) {
+    GF_ASSERT(statIdx < 40);
+    u32 value = GameStats_GetCapped(gameStats, GAME_STAT_UNK2);
+    if (value + GameStats_GetStdInc(statIdx) > 99999999) {
+        return GameStats_SetCapped(gameStats, GAME_STAT_UNK2, 99999999);
+    } else {
+        return GameStats_Add(gameStats, GAME_STAT_UNK2, GameStats_GetStdInc(statIdx));
+    }
+}
+
+u32 GameStats_GetStat2(GAME_STATS *gameStats) {
+    return GameStats_GetCapped(gameStats, GAME_STAT_UNK2);
+}
+
+void GameStats_IncSpeciesCaught(GAME_STATS *gameStats, const Pokedex *pokedex, u16 species) {
+    if (!Pokedex_CheckMonCaughtFlag(pokedex, species)) {
+        GameStats_AddSpecial(gameStats, 21);
```

Comment:
Use consts for stat index (GAME_STAT_21, which looks like it's GAME_STAT_SPECIES_CAUGHT)